core/drivers/clk/clk-stm32-core.c

```diff
@@ -0,0 +1,354 @@
+/* STM32 GATE READY CLOCK OPS */
+static TEE_Result clk_stm32_gate_ready_endisable(uint16_t gate_id, bool enable,
+						 bool wait_rdy)
+{
+	clk_stm32_endisable_gate(gate_id, enable);
+
+	if (wait_rdy)
+		return clk_stm32_wait_ready_gate(gate_id + 1, enable);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_stm32_gate_rdy_enable(uint16_t gate_id)
+{
+	return clk_stm32_gate_ready_endisable(gate_id, true, true);
+}
+
+TEE_Result clk_stm32_gate_rdy_disable(uint16_t gate_id)
+{
+	return clk_stm32_gate_ready_endisable(gate_id, false, true);
+}
+
+/* STM32 DIV API */
+#define clk_div_mask(_width) GENMASK_32(((_width) - 1U), 0U)
+
+static unsigned int _get_table_div(const struct div_table_cfg *table,
+				   unsigned int val)
+{
+	const struct div_table_cfg *clkt = NULL;
+
+	for (clkt = table; clkt->div; clkt++)
+		if (clkt->val == val)
+			return clkt->div;
+
+	return 0;
+}
+
+static unsigned int _get_table_val(const struct div_table_cfg *table,
+				   unsigned int div)
+{
+	const struct div_table_cfg *clkt = NULL;
+
+	for (clkt = table; clkt->div; clkt++)
+		if (clkt->div == div)
+			return clkt->val;
+
+	return 0;
+}
+
+static unsigned int _get_div(const struct div_table_cfg *table,
+			     unsigned int val, unsigned long flags,
+			     uint8_t width)
+{
+	if (flags & CLK_DIVIDER_ONE_BASED)
+		return val;
+
+	if (flags & CLK_DIVIDER_POWER_OF_TWO)
+		return BIT(val);
+
+	if (flags & CLK_DIVIDER_MAX_AT_ZERO)
+		return (val != 0U) ? val : BIT(width);
+
+	if (table)
+		return _get_table_div(table, val);
+
+	return val + 1U;
+}
+
+static unsigned int _get_val(const struct div_table_cfg *table,
+			     unsigned int div, unsigned long flags,
+			     uint8_t width)
+{
+	if (flags & CLK_DIVIDER_ONE_BASED)
+		return div;
+
+	if (flags & CLK_DIVIDER_POWER_OF_TWO)
+		return __builtin_ffs(div) - 1;
+
+	if (flags & CLK_DIVIDER_MAX_AT_ZERO)
+		return (div != 0U) ? div : BIT(width);
+
+	if (table)
+		return _get_table_val(table, div);
+
+	return div - 1U;
+}
+
+static bool is_power_of_2(unsigned long n)
+{
+	return (n != 0 && ((n & (n - 1)) == 0));
+}
+
+static bool _is_valid_table_div(const struct div_table_cfg *table,
+				unsigned int div)
+{
+	const struct div_table_cfg *clkt = NULL;
+
+	for (clkt = table; clkt->div; clkt++)
+		if (clkt->div == div)
+			return true;
+
+	return false;
+}
+
+static bool _is_valid_div(const struct div_table_cfg *table,
+			  unsigned int div, unsigned long flags)
+{
+	if (flags & CLK_DIVIDER_POWER_OF_TWO)
+		return is_power_of_2(div);
+
+	if (table)
+		return _is_valid_table_div(table, div);
+
+	return true;
+}
+
+static int divider_get_val(unsigned long rate, unsigned long parent_rate,
+			   const struct div_table_cfg *table, uint8_t width,
+			   unsigned long flags)
+{
+	unsigned int div = 0U;
+	unsigned int value = 0U;
+
+	div = UDIV_ROUND_NEAREST((uint64_t)parent_rate, rate);
+
+	if (!_is_valid_div(table, div, flags))
+		return -1;
+
+	value = _get_val(table, div, flags, width);
+
+	return MIN(value, clk_div_mask(width));
```

Comment:
could use `MASK_WIDTH_SHIFT(width, 0)` (and remove macro `clk_div_mask()`)
see also line 267.